File: Example/CoreData/SyncKit/QSAppDelegate.m
//
//  QSAppDelegate.m
//  SyncKit
//
//  Created by Manuel on 07/24/2016.
//  Copyright (c) 2016 Manuel. All rights reserved.
//

#import "QSAppDelegate.h"
#import "QSCompanyTableViewController.h"
#import "QSSharedCompanyTableViewController.h"
#import "QSSettingsTableViewController.h"
@import SyncKit;

@interface QSAppDelegate ()

@property (nonatomic, strong) QSCloudKitSynchronizer *synchronizer;
@property (nonatomic, strong) QSCloudKitSynchronizer *sharedSynchronizer;

@end

@implementation QSAppDelegate

- (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions
{
    [self configureCompanyTableViewController];
    [self configureSharedCompanyVC];
    [self configureSettingsVC];
    
    return YES;
}

- (void)configureCompanyTableViewController
{
    UITabBarController *tabBarController = (UITabBarController *)self.window.rootViewController;
    UINavigationController *navController = (UINavigationController *)tabBarController.viewControllers[0];
    QSCompanyTableViewController *companyVC = (QSCompanyTableViewController *)navController.topViewController;
    if ([companyVC isKindOfClass:[QSCompanyTableViewController class]]) {
        [companyVC setManagedObjectContext:self.managedObjectContext];
        [companyVC setSynchronizer:self.synchronizer];
        [companyVC setAppDelegate:self];
    }
}

- (void)configureSharedCompanyVC
{
    UITabBarController *tabBarController = (UITabBarController *)self.window.rootViewController;
    
    UINavigationController *navController2 = (UINavigationController *)tabBarController.viewControllers[1];
    QSSharedCompanyTableViewController *sharedCompanyVC = (QSSharedCompanyTableViewController *)navController2.topViewController;
    if ([sharedCompanyVC isKindOfClass:[QSSharedCompanyTableViewController class]]) {
        [(QSSharedCompanyTableViewController *)sharedCompanyVC setSynchronizer:self.sharedSynchronizer];
    }
}

- (void)configureSettingsVC
{
    UITabBarController *tabBarController = (UITabBarController *)self.window.rootViewController;
    UINavigationController *navController3 = (UINavigationController *)tabBarController.viewControllers[2];
    QSSettingsTableViewController *settingsVC = (QSSettingsTableViewController *)navController3.topViewController;
    if ([settingsVC isKindOfClass:[QSSettingsTableViewController class]]) {
        
        settingsVC.privateSynchronizer = self.synchronizer;
        settingsVC.sharedSynchronizer = self.sharedSynchronizer;
    }
}

- (void)applicationWillResignActive:(UIApplication *)application
{
    // Sent when the application is about to move from active to inactive state. This can occur for certain types of temporary interruptions (such as an incoming phone call or SMS message) or when the user quits the application and it begins the transition to the background state.
    // Use this method to pause ongoing tasks, disable timers, and throttle down OpenGL ES frame rates. Games should use this method to pause the game.
}

- (void)applicationDidEnterBackground:(UIApplication *)application
{
    // Use this method to release shared resources, save user data, invalidate timers, and store enough application state information to restore your application to its current state in case it is terminated later.
    // If your application supports background execution, this method is called instead of applicationWillTerminate: when the user quits.
}

- (void)applicationWillEnterForeground:(UIApplication *)application
{
    // Called as part of the transition from the background to the inactive state; here you can undo many of the changes made on entering the background.
}

- (void)applicationDidBecomeActive:(UIApplication *)application
{
    // Restart any tasks that were paused (or not yet started) while the application was inactive. If the application was previously in the background, optionally refresh the user interface.
}

- (void)applicationWillTerminate:(UIApplication *)application
{
    // Called when the application is about to terminate. Save data if appropriate. See also applicationDidEnterBackground:.
}

#pragma mark - Core Data stack

@synthesize managedObjectContext = _managedObjectContext;
@synthesize sharedManagedObjectContext = _sharedManagedObjectContext;
@synthesize managedObjectModel = _managedObjectModel;
@synthesize persistentStoreCoordinator = _persistentStoreCoordinator;
@synthesize sharedPersistentStoreCoordinator = _sharedPersistentStoreCoordinator;

- (NSURL *)applicationDocumentsDirectory {
    // The directory the application uses to store the Core Data store file. This code uses a directory named "com.colourbox.cloudkittest.prueba" in the application's documents directory.
    return [[[NSFileManager defaultManager] URLsForDirectory:NSDocumentDirectory inDomains:NSUserDomainMask] lastObject];
}

- (NSManagedObjectModel *)managedObjectModel {
    // The managed object model for the application. It is a fatal error for the application not to be able to find and load its model.
    if (_managedObjectModel != nil) {
        return _managedObjectModel;
    }
    NSURL *modelURL = [[NSBundle mainBundle] URLForResource:@"QSExample" withExtension:@"momd"];
    _managedObjectModel = [[NSManagedObjectModel alloc] initWithContentsOfURL:modelURL];
    return _managedObjectModel;
}

- (NSPersistentStoreCoordinator *)persistentStoreCoordinator {
    // The persistent store coordinator for the application. This implementation creates and returns a coordinator, having added the store for the application to it.
    if (_persistentStoreCoordinator != nil) {
        return _persistentStoreCoordinator;
    }
    
    // Create the coordinator and store
    NSURL *storeURL = [[self applicationDocumentsDirectory] URLByAppendingPathComponent:@"QSExample.sqlite"];
    _persistentStoreCoordinator = [self persistentStoreCoordinatorWithStoreURL:storeURL];
    
    return _persistentStoreCoordinator;
}

- (NSPersistentStoreCoordinator *)persistentStoreCoordinatorWithStoreURL:(NSURL *)storeURL
{
    _persistentStoreCoordinator = [[NSPersistentStoreCoordinator alloc] initWithManagedObjectModel:[self managedObjectModel]];
    NSError *error = nil;
    NSString *failureReason = @"There was an error creating or loading the application's saved data.";
    if (![_persistentStoreCoordinator addPersistentStoreWithType:NSSQLiteStoreType configuration:nil URL:storeURL options:nil error:&error]) {
        // Report any error we got.
        NSMutableDictionary *dict = [NSMutableDictionary dictionary];
        dict[NSLocalizedDescriptionKey] = @"Failed to initialize the application's saved data";
        dict[NSLocalizedFailureReasonErrorKey] = failureReason;
        dict[NSUnderlyingErrorKey] = error;
        error = [NSError errorWithDomain:@"YOUR_ERROR_DOMAIN" code:9999 userInfo:dict];
        // Replace this with code to handle the error appropriately.
        // abort() causes the application to generate a crash log and terminate. You should not use this function in a shipping application, although it may be useful during development.
        NSLog(@"Unresolved error %@, %@", error, [error userInfo]);
        abort();
    }
    
    return _persistentStoreCoordinator;
}

- (NSManagedObjectContext *)managedObjectContext {
    // Returns the managed object context for the application (which is already bound to the persistent store coordinator for the application.)
    if (_managedObjectContext != nil) {
        return _managedObjectContext;
    }
    
    _managedObjectContext = [self managedObjectContextWithCoordinator:[self persistentStoreCoordinator]];
    
    return _managedObjectContext;
}

- (NSManagedObjectContext *)managedObjectContextWithCoordinator:(NSPersistentStoreCoordinator *)coordinator
{
    if (!coordinator) {
        return nil;
    }
    _managedObjectContext = [[NSManagedObjectContext alloc] initWithConcurrencyType:NSMainQueueConcurrencyType];
    [_managedObjectContext setPersistentStoreCoordinator:coordinator];
    return _managedObjectContext;
}

#pragma mark - Core Data Saving support

- (void)saveContext {
    NSManagedObjectContext *managedObjectContext = self.managedObjectContext;
    if (managedObjectContext != nil) {
        NSError *error = nil;
        if ([managedObjectContext hasChanges] && ![managedObjectContext save:&error]) {
            // Replace this implementation with code to handle the error appropriately.
            // abort() causes the application to generate a crash log and terminate. You should not use this function in a shipping application, although it may be useful during development.
            NSLog(@"Unresolved error %@, %@", error, [error userInfo]);
            abort();
        }
    }
}

#pragma mark - Core Data Synchronizer

- (QSCloudKitSynchronizer *)synchronizer
{
    if (!_synchronizer) {
        _synchronizer = [QSCloudKitSynchronizer cloudKitPrivateSynchronizerWithContainerName:@"your-container-name" managedObjectContext:self.managedObjectContext];
    }
    return _synchronizer;
}

- (QSCloudKitSynchronizer *)sharedSynchronizer
{
    if (!_sharedSynchronizer) {
        _sharedSynchronizer = [QSCloudKitSynchronizer cloudKitSharedSynchronizerWithContainerName:@"your-container-name" objectModel:self.managedObjectModel];
    }
    return  _sharedSynchronizer;
}

#pragma mark - Accepting Shares

- (void)application:(UIApplication *)application userDidAcceptCloudKitShareWithMetadata:(CKShareMetadata *)cloudKitShareMetadata
{
    CKContainer *container = [CKContainer containerWithIdentifier:cloudKitShareMetadata.containerIdentifier];
    CKAcceptSharesOperation *acceptShareOperation = [[CKAcceptSharesOperation alloc] initWithShareMetadatas:@[cloudKitShareMetadata]];
    acceptShareOperation.qualityOfService = NSQualityOfServiceUserInteractive;
    acceptShareOperation.acceptSharesCompletionBlock = ^(NSError * _Nullable operationError) {
        if (operationError) {
            UIAlertController *alertController = [UIAlertController alertControllerWithTitle:@"Error" message:@"Could not accept CloudKit share" preferredStyle:UIAlertControllerStyleAlert];
            [alertController addAction:[UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleDefault handler:nil]];
            [self.window.rootViewController presentViewController:alertController animated:YES completion:nil];
        } else {
            [self.sharedSynchronizer synchronizeWithCompletion:nil];
        }
    };
    [container addOperation:acceptShareOperation];
}

#pragma mark -

- (void)didGetChangeTokenExpiredError
{
    [self deleteAllCompanies];
    [self.synchronizer eraseLocalMetadata];
    self.synchronizer = nil;
    [self configureCompanyTableViewController];
    [self configureSettingsVC];
}

- (void)deleteAllCompanies
{
    NSArray *companies = [self.managedObjectContext executeFetchRequestWithEntityName:@"QSCompany" error:nil];
    for (NSManagedObject *object in companies) {
        [self.managedObjectContext deleteObject:object];
    }
    [self.managedObjectContext save:nil];
}

@end

File: Example/CoreData/SyncKit/QSAppDelegate.swift
//
//  QSAppDelegate.swift
//  SyncKit
//
//  Created by Jérôme Haegeli on 29.08.18.
//  Copyright © 2018 Manuel. All rights reserved.
//

import SyncKit
import UIKit
import UserNotifications

@UIApplicationMain
class QSAppDelegate: UIResponder, UIApplicationDelegate {
    
    var window: UIWindow?
    
    func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions:  [UIApplicationLaunchOptionsKey : Any]? = nil) -> Bool {
        
        checkPushNotification()
        CoreDataStack.shared.setupSynchronizer()
        CoreDataStack.shared.setupSharedSynchronizer()
        
        // Override point for customization after application launch.
        let tabBarController = window?.rootViewController as? UITabBarController
        let navController = tabBarController?.viewControllers![0] as? UINavigationController
        let companyVC = navController?.topViewController as? QSCompanyTableViewController
        if (companyVC != nil) {
            (companyVC)?.managedObjectContext = CoreDataStack.shared.persistentContainer.viewContext
            (companyVC)?.synchronizer = CoreDataStack.shared.synchronizer
        }
        let navController2 = tabBarController?.viewControllers![1] as? UINavigationController
        let sharedCompanyVC = navController2?.topViewController as? QSSharedCompanyTableViewController
        if (sharedCompanyVC != nil) {
            (sharedCompanyVC)?.synchronizer = CoreDataStack.shared.sharedSynchronizer
        }
        let navController3 = tabBarController?.viewControllers![2] as? UINavigationController
        let settingsVC = navController3?.topViewController as? QSSettingsTableViewController
        if (settingsVC != nil) {
            settingsVC?.privateSynchronizer = CoreDataStack.shared.synchronizer
            settingsVC?.sharedSynchronizer = CoreDataStack.shared.sharedSynchronizer
        }
        
        if #available(iOS 10.0, *) {
            //iOS 10.0 and greater
            UNUserNotificationCenter.current().delegate = self as? UNUserNotificationCenterDelegate
            UNUserNotificationCenter.current().requestAuthorization(options: [.badge, .sound, .alert], completionHandler: { granted, error in
                DispatchQueue.main.async {
                    if granted {
                        UIApplication.shared.registerForRemoteNotifications()
                    }
                    else {
                        //Do stuff if unsuccessful...
                    }
                }
            })
        }
        else {
            //iOS 9
            let type: UIUserNotificationType = [UIUserNotificationType.badge, UIUserNotificationType.alert, UIUserNotificationType.sound]
            let setting = UIUserNotificationSettings(types: type, categories: nil)
            UIApplication.shared.registerUserNotificationSettings(setting)
            UIApplication.shared.registerForRemoteNotifications()
        }
        
        return true
    }
    
    
    func application(_ application: UIApplication, didFailToRegisterForRemoteNotificationsWithError error: Error) {
        print("Error: \(error.localizedDescription)")
    }
    
    func application(_ application: UIApplication, didRegisterForRemoteNotificationsWithDeviceToken deviceToken: Data) {
        print("didRegisterForRemoteNotificationsWithDeviceToken: \(deviceToken)")
    }
    
    func application(_ application: UIApplication, didRegister notificationSettings: UIUserNotificationSettings) {
        print("NotificationSettings: \(notificationSettings)")
    }

    func applicationWillResignActive(_ application: UIApplication) {
        // Sent when the application is about to move from active to inactive state. This can occur for certain types of temporary interruptions (such as an incoming phone call or SMS message) or when the user quits the application and it begins the transition to the background state.
        // Use this method to pause ongoing tasks, disable timers, and throttle down OpenGL ES frame rates. Games should use this method to pause the game.
    }
    
    func applicationDidEnterBackground(_ application: UIApplication) {
        // Use this method to release shared resources, save user data, invalidate timers, and store enough application state information to restore your application to its current state in case it is terminated later.
        // If your application supports background execution, this method is called instead of applicationWillTerminate: when the user quits.
    }
    
    func applicationWillEnterForeground(_ application: UIApplication) {
        // Called as part of the transition from the background to the inactive state; here you can undo many of the changes made on entering the background.
    }
    
    func applicationDidBecomeActive(_ application: UIApplication) {
        // Restart any tasks that were paused (or not yet started) while the application was inactive. If the application was previously in the background, optionally refresh the user interface.
    }
    
    func applicationWillTerminate(_ application: UIApplication) {
        // Called when the application is about to terminate. Save data if appropriate. See also applicationDidEnterBackground:.
    }

    
    func application(_ application: UIApplication, didReceiveRemoteNotification userInfo: [AnyHashable : Any]) {
        print("Remote Notification received")
        // Detect if APN is received on Background or Foreground state
        if application.applicationState == .inactive {
            print("Application is inactive")
        } else if application.applicationState == .active {
            print("Application is active")
        }
        CoreDataStack.shared.synchronizer?.synchronize(completion: nil)
        CoreDataStack.shared.sharedSynchronizer?.synchronize(completion: nil)
    }

    
// MARK: - Accepting Shares
    func application(_ application: UIApplication, userDidAcceptCloudKitShareWith cloudKitShareMetadata: CKShareMetadata) {
        let container = CKContainer(identifier: cloudKitShareMetadata.containerIdentifier)
        let acceptShareOperation = CKAcceptSharesOperation(shareMetadatas: [cloudKitShareMetadata])
        acceptShareOperation.qualityOfService = .userInteractive
        acceptShareOperation.acceptSharesCompletionBlock = { operationError in
            if operationError != nil {
                let alertController = UIAlertController(title: "Error", message: "Could not accept CloudKit share", preferredStyle: .alert)
                alertController.addAction(UIAlertAction(title: "OK", style: .default, handler: nil))
                self.window?.rootViewController?.present(alertController, animated: true)
            } else {
                CoreDataStack.shared.sharedSynchronizer?.synchronize(completion: nil)
            }
        }
        container.add(acceptShareOperation)
    }
    
    //MARK: - Check Notifications
    
    func checkPushNotification(checkNotificationStatus isEnable : ((Bool)->())? = nil){
        if #available(iOS 10.0, *) {
            UNUserNotificationCenter.current().getNotificationSettings(){ (setttings) in
                
                switch setttings.authorizationStatus{
                case .authorized:
                    // User has given authorization.
                    print("enabled notification setting")
                    isEnable?(true)
                case .denied:
                    // User has denied authorization.
                    // You could tell them to change this in Settings
                    print("setting has been disabled")
                    isEnable?(false)
                case .notDetermined:
                    // Authorization request has not been made yet
                    print("something vital went wrong here")
                    isEnable?(false)
                }
            }
        } else {
            // iOS 9 support
            let isNotificationEnabled = UIApplication.shared.currentUserNotificationSettings?.types.contains(UIUserNotificationType.alert)
            if isNotificationEnabled == true{
                
                print("enabled notification setting")
                isEnable?(true)
                
            }else{
                
                print("setting has been disabled")
                isEnable?(false)
            }
        }
    }

    func requestNotificationAuthorization() -> Bool {
        if #available(iOS 10, *)
        { // iOS 10 support
            //create the notificationCenter
            let center = UNUserNotificationCenter.current()
            center.delegate = UIApplication.shared.delegate as? UNUserNotificationCenterDelegate
            // set the type as sound or badge
            var result = false
            center.requestAuthorization(options: [.sound,.alert,.badge]) { (granted, error) in
                if granted {
                    print("Notification Enable Successfully")
                    result = true
                }else{
                    print("Some Error Occure")
                    result = false
                }
            }
            UIApplication.shared.registerForRemoteNotifications()
            return result
        }
        else
        {
            // iOS 9 support
            UIApplication.shared.registerUserNotificationSettings(UIUserNotificationSettings(types: [.badge, .sound, .alert], categories: nil))
            UIApplication.shared.registerForRemoteNotifications()
            return true
        }
        
    }

}
